src/compiler/csharp_generator.cc

```diff
@@ -323,17 +323,85 @@ std::vector<const Descriptor*> GetUsedMessages(
   return result;
 }
 
-void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service) {
+void GenerateMarshallerFields(Printer* out, const ServiceDescriptor* service,
+                              bool use_buffer_serialization) {
   std::vector<const Descriptor*> used_messages = GetUsedMessages(service);
-  for (size_t i = 0; i < used_messages.size(); i++) {
-    const Descriptor* message = used_messages[i];
+  if (used_messages.size() == 0)
+  {
+    out->Print("\n");
+    return;
+  }
+
+  if (use_buffer_serialization) {
+    // Generate buffer serialization marshallers
+    out->Print(
+        "static void __Helper_SerializeMessage("
+        "global::Google.Protobuf.IMessage message, "
+        "grpc::SerializationContext context)\n"
+        "{\n");
+    out->Indent();
+    out->Print(
+          "#if !GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION\n"
+          "var bufferMessage = message as global::Google.Protobuf.IBufferMessage;\n"
+          "if (bufferMessage != null)\n"
+          "{\n");
+    out->Indent();
+    out->Print(
+            "var writer = new global::Google.Protobuf.CodedOutputWriter("
```

Comment:
Add `context.SetPayloadLength(bufferMessage.CalculateSize());` above this line.

Setting the length before the buffer is accessed allows the message header to be written immediately, and the message content is written directly to the response writer.

If we don't set it then the message content has to be buffered.